File: solutions/cpp/Completion_search.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <iterator>
#include <numeric>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::ostream_iterator;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
double CompletionSearch(double budget, vector<double>* A) {
  sort(A->begin(), A->end());
  // Calculates the prefix sum for A.
  vector<double> prefix_sum;
  partial_sum(A->cbegin(), A->cend(), back_inserter(prefix_sum));
  // costs[i] represents the total payroll if the cap is A[i].
  vector<double> costs;
  for (int i = 0; i < prefix_sum.size(); ++i) {
    costs.emplace_back(prefix_sum[i] + (A->size() - i - 1) * (*A)[i]);
  }

  auto lower = lower_bound(costs.cbegin(), costs.cend(), budget);
  if (lower == costs.cend()) {
    return -1.0;  // No solution since budget is too large.
  }

  if (lower == costs.cbegin()) {
    return budget / A->size();
  }
  auto idx = distance(costs.cbegin(), lower);
  return (budget - prefix_sum[idx - 1]) / (A->size() - idx);
}
// @exclude

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 10000; ++times) {
    int n;
    vector<double> A;
    double tar;
    if (argc == 2) {
      n = atoi(argv[1]);
      uniform_int_distribution<int> dis(0, 99999);
      tar = dis(gen);
    } else if (argc == 3) {
      n = atoi(argv[1]), tar = atoi(argv[2]);
    } else {
      uniform_int_distribution<int> n_dis(1, 1000);
      n = n_dis(gen);
      uniform_int_distribution<int> tar_dis(0, 99999);
      tar = tar_dis(gen);
    }
    for (int i = 0; i < n; ++i) {
      uniform_int_distribution<int> dis(0, 9999);
      A.emplace_back(dis(gen));
    }
    cout << "A = ";
    copy(A.begin(), A.end(), ostream_iterator<double>(cout, " "));
    cout << endl;
    cout << "tar = " << tar << endl;
    double ret = CompletionSearch(tar, &A);
    if (ret != -1.0) {
      cout << "ret = " << ret << endl;
      double sum = 0.0;
      for (int i = 0; i < n; ++i) {
        if (A[i] > ret) {
          sum += ret;
        } else {
          sum += A[i];
        }
      }
      tar -= sum;
      cout << "sum = " << sum << endl;
      assert(tar < 1.0e-8);
    }
  }
  return 0;
}

File: solutions/python/completion_search.py
import random
import sys


# @include
def find_salary_cap(target_payroll, current_salaries):
    current_salaries.sort()
    unadjusted_salary_sum = 0.0
    for i, current_salary in enumerate(current_salaries):
        adjusted_people = len(current_salaries) - i
        adjusted_salary_sum = current_salary * adjusted_people
        if unadjusted_salary_sum + adjusted_salary_sum >= target_payroll:
            return (target_payroll - unadjusted_salary_sum) / adjusted_people
        unadjusted_salary_sum += current_salary
    # No solution, since target_payroll > existing payroll.
    return -1.0
# @exclude


def small_test():
    A = [20, 30, 40, 90, 100]
    T = 210
    assert find_salary_cap(T, A) == 60
    T = 280
    assert find_salary_cap(T, A) == 100
    T = 50
    assert find_salary_cap(T, A) == 10
    T = 281
    assert find_salary_cap(T, A) == -1.0


def main():
    small_test()
    for _ in range(10000):
        if len(sys.argv) == 2:
            n = int(sys.argv[1])
            tar = random.randrange(100000)
        elif len(sys.argv) == 3:
            n = int(sys.argv[1])
            tar = int(sys.argv[2])
        else:
            n = random.randint(1, 1000)
            tar = random.randrange(100000)
        A = [random.randrange(10000) for _ in range(n)]
        print('A = ', end='')
        print(*A)
        print('tar =', tar)
        ret = find_salary_cap(tar, A)
        A.sort()
        if ret != -1.0:
            print('ret =', ret)
            sum = 0.0
            for i in range(n):
                if A[i] > ret:
                    sum += ret
                else:
                    sum += A[i]
            tar -= sum
            print('sum =', sum)
            assert tar < 1.0e-8


if __name__ == '__main__':
    main()
